add shellcheck and fix complaints

tuminoid opened the pull request:
Add shellcheck. Fix all the shellcheck complaints, and make shell code consistent Bash.
    
Another PR in project-infra will activate the check for PRs.

Markdownlint is separated to its own PR #413

tuminoid commented:
/test-ubuntu-integration-main

elfosardo commented on prepare-efi.sh:
```diff
@@ -28,17 +28,17 @@ fi
 # ``Cannot initialize '::'``
 # This is due to the conversion table missing codepage 850, included in glibc-gconv-extra
-dnf install -y grub2 shim dosfstools mtools glibc-gconv-extra $PACKAGES
+dnf install -y grub2 shim dosfstools mtools glibc-gconv-extra "${PACKAGES}"
```
maybe add disable=SC2086 here?
the PACKAGES list will be considered as a single string, afraid this will break the build

tuminoid replied:
PACKAGES is defined above to be a single word `grub2-efi-x86`, so it will not break. Variable naming is misleading.

elfosardo commented on prepare-image.sh:
```diff
@@ -11,16 +11,16 @@ echo "tsflags=nodocs" >> /etc/dnf/dnf.conf
     dnf config-manager --set-enabled crb
     dnf upgrade -y
-    xargs -rtd'\n' dnf install -y < $IRONIC_PKG_LIST
+    xargs -rtd'\n' dnf install -y < "${IRONIC_PKG_LIST}"
```
same here, probably better add disable=SC2086

tuminoid replied:
IRONIC_PKG_LIST is a file name and its read by `<`, so here quotes are correct.

elfosardo commented on prepare-image.sh:
```diff
@@ -29,38 +29,37 @@ if [[ $INSTALL_TYPE == "source" ]]; then
     dnf upgrade -y
     # NOTE(dtantsur): pip is a requirement of python3 in CentOS
-    dnf install -y python3-pip $BUILD_DEPS
+    dnf install -y python3-pip "${BUILD_DEPS}"
```
ditto

elfosardo commented on prepare-image.sh:
```diff
@@ -29,38 +29,37 @@ if [[ $INSTALL_TYPE == "source" ]]; then
 
     # clean installed build dependencies
-    dnf remove -y $BUILD_DEPS
+    dnf remove -y "${BUILD_DEPS}"
```
ditto

tuminoid replied:
I'll fix that.

tuminoid commented:
/test-ubuntu-integration-main

tuminoid commented:
/test-centos-integration-main-ironic-source

elfosardo commented:
/approve

dtantsur reviewed (changes requested):
I absolutely like the markdown verification, but the diff in the scripts seems mostly useless. There is no point to enforce ${variables} over $variables except that in a few cases.

Landing this will also turn our life downstream into a nightmare, because no backports will apply cleanly (if you do backporting, you'll have the same problem too).

tuminoid commented on scripts/configure-httpd-ipa.sh:
```diff
@@ -1,23 +1,24 @@
 
 INSPECTOR_EXTRA_ARGS=" ipa-inspection-callback-url=${IRONIC_BASE_URL}:${IRONIC_INSPECTOR_ACCESS_PORT}/v1/continue"
-if [[ $IRONIC_FAST_TRACK == true ]]; then
+if [[ $IRONIC_FAST_TRACK = true ]]; then
```
`==` and `=` are effecively the same thing, except `==` is Bash-only and `=` is POSIX compatible.

That said, the usage of `=` and `==` is  very mixed in the scripts, and I just went consistently for `=`. I can change the other way around, if we end up taking this PR in.

tuminoid replied:
Fixed all of the `=` to `==` as requested. Also made all `[ .. ]` to `[[ .. ]]` as `[ ` is POSIX and `==` is Bash-only and their combination is not supported (yes, Bash doesn't explode, but officially its undefined combination).

dtantsur commented:
As a note, the Google style guide is pretty opinionated and may not reflect how most people write scripts. Nor does it reflect the intention of $ vs ${}, which is: ${} is used when the variable name is followed by characters that are valid in a variable name, e.g. `${foo}_bar` (cannot be `$foo_bar`).

tuminoid commented:
> > Stay consistent with what you find

There is no consistency like I said already, so nothing to stick by.

tuminoid commented:
/hold

Holding this so it doesn't get merged by drive-by lgtm. @elfosardo and @dtantsur please settle if you are willing to take this in. I'm happy to make adjustments if required, but as stated before, I'm going for consistency, maintainability and readability, preferably organization wide. If this does not work for you, I can drop this, and you can implement it in a way that works for you.

tuminoid commented:
/test-ubuntu-integration-main 
/test-centos-integration-main-ironic-source

tuminoid commented:
/test-ubuntu-integration-main
/test-centos-integration-main-ironic-source

tuminoid commented:
@dtantsur @elfosardo Its been a week. Have you guys discussed this?

tuminoid commented:
> @dtantsur @elfosardo Its been a week. Have you guys discussed this?

It's now been two weeks. Can you guys discuss your mutual stance on this?

The question in the end is rather simple: Do you want to take the PR in or not? We have bunch of shell code that is a bit messy, not linted, not consistent. We have here PR cleaning it up for you, maybe a bit opinionated PR but guess what, most of the languages/ecossytems have rather opinionated best practices/style guides these days (gofmt, prettier, k8s and so on), because its effective. They're easy for anyone to work in, you can automate your editors and your CI.

> I have no visibiility into your internal development practices, and in this repo there are no branches. I'm trying to have all repositories under metal3-io to have consistent coding style, basic quality checks in place etc so they are in the long run easier to maintain, developers have safety nets in place and faster feedback cycle.
> 
> If you feel this change is not worth of merging or causes too much trouble for you, then it is your right to decline it, but I'd hope the opposite, that you possibly take this into your backports, and after that there is no more conflicts. But like I said, I got no idea what your setup looks like and if that is feasible.

This still stands as well.

elfosardo commented:
/test-centos-integration-main-ironic-source

tuminoid commented:
Took away all curly brace changes etc. @dtantsur PTAL.

/test-ubuntu-integration-main

tuminoid commented:
/test-centos-integration-main-ironic-source

tuminoid commented:
/test-centos-integration-main-ironic-source
/test-ubuntu-integration-main

tuminoid commented:
/test-centos-integration-main-ironic-source
/test-ubuntu-integration-main

dtantsur commented on scripts/coreos-ipa-common.sh:
```diff
@@ -4,18 +4,24 @@ ROOTFS_FILE=${ROOTFS_FILE:-/shared/html/images/ironic-python-agent.rootfs}
 if use_coreos_ipa; then
-    export IRONIC_KERNEL_PARAMS="${IRONIC_KERNEL_PARAMS:-} $(coreos_kernel_params)"
+    IRONIC_KERNEL_PARAMS="${IRONIC_KERNEL_PARAMS:-} $(coreos_kernel_params)"
+    export IRONIC_KERNEL_PARAMS
```
I wonder why the split is required...

tuminoid replied:
`$(coreos_kernel_params)` is subshell call and exporting it on the same line masks the possible error from it.

dtantsur commented on scripts/runhttpd:
```diff
@@ -1,16 +1,18 @@
 #!/usr/bin/bash
 
+# shellcheck disable=SC1091
```
Cannot we follow the recommendation in https://www.shellcheck.net/wiki/SC1090 instead of disabling it?

tuminoid replied:
`disable=SC1091` needs to be there, since shellcheck won't follow `source=xxx` as `/bin/xxx` is considered external source by shellcheck. We could use `SC1090` recommendation, if it was pointing to non-absolute path.

dtantsur commented on scripts/runhttpd:
```diff
@@ -23,28 +25,29 @@ chmod 0777 /shared/html
 fi
 export INSPECTOR_EXTRA_ARGS
 
+# shellcheck disable=SC1091
```
same

tuminoid replied:
Same as above.

dtantsur commented:
Much better, thanks! I'm not a fan of some changes, but I understand the consistency argument here. Just one question inline.

elfosardo commented on scripts/configure-ironic.sh:
```diff
@@ -11,16 +11,19 @@ IRONIC_EXTERNAL_IP="${IRONIC_EXTERNAL_IP:-}"
 #   <interface.vlan> - a particular VLAN on an interface, not relying on LLDP
 export IRONIC_INSPECTOR_VLAN_INTERFACES=${IRONIC_INSPECTOR_VLAN_INTERFACES:-all}
 
+# shellcheck disable=SC1091
```
you could use curly brackets to group multiple lines and remove the second disable statement, I'm ok either way

tuminoid replied:
True. IMO it is easier to understand this way, while being explicit about the ignores.

elfosardo commented on scripts/configure-ironic.sh:
```diff
@@ -40,17 +43,18 @@ wait_for_interface_or_ip
-    if [ "$IRONIC_VMEDIA_TLS_SETUP" = "true" ]; then
+if [[ -n "$IRONIC_EXTERNAL_IP" ]]; then
+    IRONIC_EXTERNAL_CALLBACK_URL="${IRONIC_SCHEME}://${IRONIC_EXTERNAL_IP}:${IRONIC_ACCESS_PORT}"
+    export IRONIC_EXTERNAL_CALLBACK_URL
```
is the split necessary here?

tuminoid replied:
No its not, good catch. Fixed.

tuminoid commented:
I took some more unnecessary changes off, and I think I addressed all comments.

/test-centos-integration-main-ironic-source
/test-ubuntu-integration-main

tuminoid commented:
Cleura is busted but one more try for today.

/test-centos-integration-main-ironic-source
/test-ubuntu-integration-main

tuminoid commented:
/cc @dtantsur 

Good for LGTM now?

lentzi90 reviewed (commented):
/lgtm

tuminoid commented:
@dtantsur can you PTAL and remove changes requested so we can merge this? All comments should be addressed now.
